src/zns/utils.ts

```diff
@@ -0,0 +1,264 @@
+      ret.push((acc << (toWidth - bits)) & maxv);
+    }
+  } else if (bits >= fromWidth || (acc << (toWidth - bits)) & maxv) {
+    return null;
+  }
+
+  return Buffer.from(ret);
+};
+
+function hrpExpand (hrp: string): Buffer {
+  const ret = [];
+  let p;
+  for (p = 0; p < hrp.length; ++p) {
+    ret.push(hrp.charCodeAt(p) >> 5);
+  }
+  ret.push(0);
+  for (p = 0; p < hrp.length; ++p) {
+    ret.push(hrp.charCodeAt(p) & 31);
+  }
+  return Buffer.from(ret);
+};
+
+function polymod (values: Buffer): number {
+  let chk = 1;
+  // tslint:disable-next-line
+  for (let p = 0; p < values.length; ++p) {
+    const top = chk >> 25;
+    chk = ((chk & 0x1ffffff) << 5) ^ values[p];
+    for (let i = 0; i < 5; ++i) {
+      if ((top >> i) & 1) {
+        chk ^= GENERATOR[i];
+      }
+    }
+  }
+  return chk;
+};
+
+function createChecksum(hrp: string, data: Buffer) {
+  const values = Buffer.concat([
+    Buffer.from(hrpExpand(hrp)),
+    data,
+    Buffer.from([0, 0, 0, 0, 0, 0]),
+  ]);
+  // var values = hrpExpand(hrp).concat(data).concat([0, 0, 0, 0, 0, 0]);
+  const mod = polymod(values) ^ 1;
+  const ret = [];
+  for (let p = 0; p < 6; ++p) {
+    ret.push((mod >> (5 * (5 - p))) & 31);
+  }
+  return Buffer.from(ret);
+}
+
+
+function verifyChecksum(hrp: string, data: Buffer) {
+  return polymod(Buffer.concat([hrpExpand(hrp), data])) === 1;
+}
+
+function encode(hrp: string, data: Buffer) {
+  const combined = Buffer.concat([data, createChecksum(hrp, data)]);
+  let ret = hrp + '1';
+  // tslint:disable-next-line
+  for (let p = 0; p < combined.length; ++p) {
+    ret += CHARSET.charAt(combined[p]);
+  }
+  return ret;
+};
+
+function decode (bechString: string) {
+  let p;
+  let hasLower = false;
+  let hasUpper = false;
+  for (p = 0; p < bechString.length; ++p) {
+    if (bechString.charCodeAt(p) < 33 || bechString.charCodeAt(p) > 126) {
+      return null;
+    }
+    if (bechString.charCodeAt(p) >= 97 && bechString.charCodeAt(p) <= 122) {
+      hasLower = true;
+    }
+    if (bechString.charCodeAt(p) >= 65 && bechString.charCodeAt(p) <= 90) {
+      hasUpper = true;
+    }
+  }
+  if (hasLower && hasUpper) {
+    return null;
+  }
+  bechString = bechString.toLowerCase();
+  const pos = bechString.lastIndexOf('1');
+  if (pos < 1 || pos + 7 > bechString.length || bechString.length > 90) {
+    return null;
+  }
+  const hrp = bechString.substring(0, pos);
+  const data = [];
+  for (p = pos + 1; p < bechString.length; ++p) {
+    const d = CHARSET.indexOf(bechString.charAt(p));
+    if (d === -1) {
+      return null;
+    }
+    data.push(d);
+  }
+
+  if (!verifyChecksum(hrp, Buffer.from(data))) {
+    return null;
+  }
+
+  return { hrp, data: Buffer.from(data.slice(0, data.length - 6)) };
+};
+
+/**
+ * toChecksumAddress
+ *
+ * takes hex-encoded string and returns the corresponding address
+ *
+ * @param {string} address
+ * @returns {string}
+ */
+export const toChecksumAddress = (address: string): string => {
```

Comment:
What is the motivation to make it arrow function instead of "normal" function?

I also believe that we shouldn't use `@param` and `@returns` tags that don't bring any new info in the comments (we already have the info about the types in .d.ts files, don't we)?